### PR TITLE
Adding Concurrency in Pool Manager

### PR DESCRIPTION
**Pull request by therahulbhati**
- Removed router side caching
- Introduced concurrency in pool manager, so that multiple function pods can specialize and serve request concurrently
- Added concurrency field in fission CLI, using which one can define concurrency while creating functions. Default to 5.

Signed-off-by: therahulbhati <rjbhati009@gmail.com>

---
This change is [Reviewable](https://reviewable.io/reviews/fission/fission/1698)

### CODE REVIEW

**Comment by vishal-biyani** [comment]
Also Fixes https://github.com/fission/fission/issues/1628

**Comment by vishal-biyani** [comment]
When the concurrency limit is hit and a user tries to run `fission fn test`, the error message is not very helpful to end-user.

```
Error: Error calling function another: 502; Please try again or fix the error: error sending request to function

Error: error getting function response
```

File: pkg/apis/core/v1/types.go [hunk 1]
```diff
@@ -349,6 +349,10 @@ type (
 		// is detected within the idle timeout, the executor will then recycle the
 		// function pod(s) to release resources.
 		IdleTimeout *int `json:"idletimeout,omitempty"`
+
+		// Maximum number of concurrent requests to be served
```
Comment (vishal-biyani) [inline pkg/apis/core/v1/types.go:353]:
```suggestion
		// Maximum number of pods to be specialized which will serve requests
```

File: pkg/executor/api.go [hunk 2]
```diff
@@ -61,7 +62,29 @@ func (executor *Executor) getServiceForFunctionApi(w http.ResponseWriter, r *htt
 		return
 	}
 
-	serviceName, err := executor.getServiceForFunction(fn)
+	t := fn.Spec.InvokeStrategy.ExecutionStrategy.ExecutorType
```
Comment (vishal-biyani) [inline pkg/executor/api.go:65]: I noticed that we are getting function from K8S API - which is in the request path - should we raise an issue for improving that performance?

File: pkg/executor/api.go [hunk 3]
```diff
@@ -84,18 +107,12 @@ func (executor *Executor) getServiceForFunctionApi(w http.ResponseWriter, r *htt
 // stale addresses are not returned to the router.
 // To make it optimal, plan is to add an eager cache invalidator function that watches for pod deletion events and
 // invalidates the cache entry if the pod address was cached.
-func (executor *Executor) getServiceForFunction(fn *fv1.Function) (string, error) {
+func (executor *Executor) getServiceForFunction(fn *fv1.Function, et executortype.ExecutorType) (string, error) {
```
Comment (vishal-biyani) [inline pkg/executor/api.go:110]: Do we really need to pass the executor type? I think we have it in the function itself right?

File: pkg/executor/api.go [hunk 4]
```diff
@@ -181,12 +198,50 @@ func (executor *Executor) healthHandler(w http.ResponseWriter, r *http.Request)
...
+		return
+	}
+	tapSvcReq := client.TapServiceRequest{}
+	err = json.Unmarshal(body, &tapSvcReq)
+	if err != nil {
+		http.Error(w, "Failed to parse request", http.StatusBadRequest)
+		return
+	}
+
+	fn, err := executor.fissionClient.CoreV1().Functions(tapSvcReq.FnMetadata.Namespace).Get(tapSvcReq.FnMetadata.Name, metav1.GetOptions{})
```
Comment (vishal-biyani) [inline pkg/executor/api.go:214]: I think we will need to optimize this in another PR similar to the one from `getServiceForFunctionApi`

File: pkg/executor/executortype/newdeploy/newdeploymgr.go [hunk 5]
```diff
@@ -153,6 +153,14 @@ func (deploy *NewDeploy) DeleteFuncSvcFromCache(fsvc *fscache.FuncSvc) {
...
 }
 
+func (deploy *NewDeploy) UnTapService(fn *fv1.Function, svcHost string) {
+	// TO DO
+}
+
+func (deploy *NewDeploy) GetActiveInstances(fn *fv1.Function) int {
+	return 0
+}
+
```
Comment (vishal-biyani) [inline pkg/executor/executortype/newdeploy/newdeploymgr.go:163]: Can you please leave a detailed comment :)

File: pkg/executor/executortype/poolmgr/gpm.go [hunk 6]
```diff
@@ -166,11 +166,19 @@ func (gpm *GenericPoolManager) GetFuncSvc(ctx context.Context, fn *fv1.Function)
 }
 
 func (gpm *GenericPoolManager) GetFuncSvcFromCache(fn *fv1.Function) (*fscache.FuncSvc, error) {
-	return gpm.fsCache.GetByFunction(&fn.ObjectMeta)
+	return gpm.fsCache.GetConnFunction(&fn.ObjectMeta)
```
Comment (vishal-biyani) [inline pkg/executor/executortype/poolmgr/gpm.go:169]: What does `GetConnFunction` mean - or can we use a better name, please?

File: pkg/executor/fscache/functionServiceCache.go [hunk 7]
```diff
@@ -173,6 +205,23 @@ func (fsc *FunctionServiceCache) GetByFunctionUID(uid types.UID) (*FuncSvc, erro
...
+	fsc.setFuncAlive(fsvc.Function.Name, string(fsvc.Function.UID), true)
+}
+
+func (fsc *FunctionServiceCache) GetActiveInstances(m *metav1.ObjectMeta) int {
+	return fsc.connFunctionCache.GetTotalActive(crd.CacheKey(m))
+}
+
+func (fsc *FunctionServiceCache) SetInActive(fn *fv1.Function, svcHost string) {
+	fsc.connFunctionCache.UnSet(crd.CacheKey(&fn.ObjectMeta), svcHost)
+}
```
Comment (vishal-biyani) [inline pkg/executor/fscache/functionServiceCache.go:223]: Let's use `Active` --> `Available` in all places?

File: pkg/fission-cli/flag/flag.go [hunk 8]
```diff
@@ -110,6 +110,7 @@ var (
 	FnTestHeader            = Flag{Type: StringSlice, Name: flagkey.FnTestHeader, Short: "H", Usage: "Request headers"}
 	FnTestQuery             = Flag{Type: StringSlice, Name: flagkey.FnTestQuery, Short: "q", Usage: "Request query parameters: -q key1=value1 -q key2=value2"}
 	FnIdleTimeout           = Flag{Type: Int, Name: flagkey.FnIdleTimeout, Usage: "The length of time (in seconds) that a function is idle before pod(s) are eligible for recycling", DefaultValue: 120}
+	FnConcurrency           = Flag{Type: Int, Name: flagkey.FnConcurrency, Aliases: []string{"con"}, Usage: "Maximum number of concurrent requests to be served", DefaultValue: 5}
```
Comment (vishal-biyani) [inline pkg/fission-cli/flag/flag.go:113]:
```suggestion
	FnConcurrency           = Flag{Type: Int, Name: flagkey.FnConcurrency, Aliases: []string{"con"}, Usage: "Maximum number of pods specialized concurrently to serve requests", DefaultValue: 5}
```

File: pkg/newcache/newcache.go [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package newcache implements a simple cache implementation having one value mapped by two keys.
+// As of now this package is only used by poolmanager executor
+package newcache
```
Comment (vishal-biyani) [inline pkg/newcache/newcache.go:19]: Can we use a better name instead of `newcache` please?

File: pkg/router/functionHandler_test.go [hunk 10]
```diff
@@ -34,71 +32,71 @@ import (
...
+// 			timeoutExponent: 2,
+// 			maxRetries:      10,
+// 		},
+// 		httpTrigger: httpTrigger,
+// 	}
+// 	functionHandlerServer := httptest.NewServer(http.HandlerFunc(fh.handler))
+// 	fhURL := functionHandlerServer.URL
+
+// 	testRequest(fhURL, testResponseString)
+// }
```
Comment (vishal-biyani) [inline pkg/router/functionHandler_test.go:99]: Please remove or uncomment as appropriate.

File: pkg/router/router_test.go [hunk 11]
```diff
@@ -16,93 +16,79 @@ limitations under the License.
...
-	testRequest(testUrl, testResponseString)
-}
+// func TestRouter(t *testing.T) {
+// 	// metadata for a fake function
+// 	fnMeta := metav1.ObjectMeta{Name: "foo", Namespace: metav1.NamespaceDefault}
+
+// 	// and a reference to it
+// 	fr := fv1.FunctionReference{
+// 		Type: fv1.FunctionReferenceTypeFunctionName,
+// 		Name: fnMeta.Name,
```
Comment (vishal-biyani) [inline pkg/router/router_test.go:26]: Please uncomment or remove as appropriate.

**Review by vishal-biyani — CHANGES_REQUESTED** [review]

File: pkg/executor/api.go (hunk 3 above)
Comment (therahulbhati) [reply pkg/executor/api.go:110]: Done

File: pkg/executor/executortype/newdeploy/newdeploymgr.go (hunk 5 above)
Comment (therahulbhati) [reply pkg/executor/executortype/newdeploy/newdeploymgr.go]: Done

File: pkg/newcache/newcache.go (hunk 9 above)
Comment (therahulbhati) [reply pkg/newcache/newcache.go]: Done

File: pkg/executor/fscache/functionServiceCache.go (hunk 7 above)
Comment (therahulbhati) [reply pkg/executor/fscache/functionServiceCache.go]: Done

File: pkg/executor/api.go (hunk 4 above)
Comment (therahulbhati) [reply pkg/executor/api.go:216]: https://github.com/fission/fission/issues/1713

File: pkg/executor/api.go (hunk 2 above)
Comment (therahulbhati) [reply pkg/executor/api.go:64]: https://github.com/fission/fission/issues/1713

File: pkg/router/functionHandler_test.go (hunk 10 above)
Comment (therahulbhati) [reply pkg/router/functionHandler_test.go]: Done

File: pkg/router/router_test.go (hunk 11 above)
Comment (therahulbhati) [reply pkg/router/router_test.go]: Done

File: pkg/executor/executortype/poolmgr/gpm.go (hunk 6 above)
Comment (therahulbhati) [reply pkg/executor/executortype/poolmgr/gpm.go]: Done

**Review by vishal-biyani — APPROVED** [review]

**Comment by vishal-biyani** [comment]
Thanks, @therahulbhati 🎉 This is a very useful feature for users.

**Pull request merged**